polynote-server/src/main/scala/polynote/server/SharedNotebook.scala

```diff
@@ -214,11 +216,12 @@ class IOSharedNotebook(
       }
     }(_ => kernelLock.release)
 
+    // we want to queue all the cells, but evaluate them in order. So the outer IO of the result runs the outer IO of queueCell for all the cells.
     override def runCells(ids: List[CellID]): IO[Stream[IO, CellResult]] =
-      ensureKernel().map {
-        kernel => Stream.emits(ids).evalMap {
-          id => runCell(id).map(results => results.map(result => CellResult(ShortString(path), id, result)))
-        }.flatten
+      ids.map {
+        id => queueCell(id).map(_.map(_.map(result => CellResult(ShortString(path), id, result))))
```

Comment:
No fancy operator for `.map(_.map(_.map(`? 😛

Reply:
Could use a nested functor instance, but I doubt it would look nicer :)